### PR TITLE
(feat) infer slot/events type from props

### PR DESCRIPTION
**Pull request by dummdidumm**
This changes the generation for accessing the events/slots of a component. Instead of using `instanceOf`, the component instance is created using the constructor and passing the props. This enables TypeScript to infer generic relationships between props and slots/events.
#945
Note that this change does not cover all cases: If someone forwards a slot or an event whose type is generic and needs to be infered, that type will remain unknown, because reinstantiating the context in the `return { props: .. }` step is really hard.

TODO:
How to fix the "varialbe used before instantiation"-error that occurs when we tranform `<Foo prop={prop} let:prop>`? The transformation for `let:prop` is `const prop = new Foo({.. props: {prop: prop}})` which TS doesn't like. @jasonlyu123 any ideas?

### CODE REVIEW

**Comment by jasonlyu123** [comment]
The only solution I can think of now is to declare a new variable. Something similar to if-scope or pass it as a parameter to an immediately invoked function.
```ts
const __propsArgs = { a }
() =>{  
  const a = new Component({ props: __propsArgs  }).$$slotdef['default']
}
```
or

```ts
((__propsArgs) =>{  
  const a = new Component({ props: __propsArgs  }).$$slotdef['default']
})({ a })
```

**Comment by dummdidumm** [comment]
Yeah sounds like the only option. So there needs to be some logic to track which props there are both on the default and the names slots, and if there's one that's duplicated, construct a immediately invoked function from the default slot and reuse that in all the inner slots.

**Comment by dummdidumm** [comment]
Ok I reused the ifScope logic here, worked out well, although I still feel that this code is getting more and more complex - if it's just inevitable or I'm writing shitty code, can't tell for sure right now 😄 @jasonlyu123 could you have a look?

**Review by jasonlyu123 — CHANGES_REQUESTED** [review]
Mostly working great. Found a small problem when an attribute is empty. And some minor issues to discuss.

File: packages/svelte2tsx/src/htmlxtojsx/utils/node-utils.ts [hunk 1]
```diff
@@ -9,6 +10,152 @@ export function getTypeForComponent(node: Node): string {
...
+                if (val.type === 'MustacheTag') {
+                    const valueStr = originalStr.substring(val.start + 1, val.end - 1);
+                    if (val.expression.type === 'Identifier') {
+                        return { name, value: valueStr, identifier: valueStr };
+                    }
+                    return { name, value: valueStr, identifier: true };
+                }
+            }
+
+            const value =
```
[the remark this answers is not in the document]
Comment (dummdidumm) [reply packages/svelte2tsx/src/htmlxtojsx/utils/node-utils.ts:134]: Damn, you are completely right, this is just a string in the end and we can simplify this!

**Comment by dummdidumm** [comment]
These were great suggestions, thanks! Especially that string template comment was a moment of enlightment 😄

**Comment by ryu-man** [comment]
I noticed that a string value is inferred as "any" type in the slot prop!

**Comment by dummdidumm** [comment]
Should be fixed with the latest release

**Comment by ryu-man** [comment]
I am using the latest release, all values get inferred normally, but when using a string value it inferred as "any" type not "string" type

**Comment by dummdidumm** [comment]
Can't reproduce. What's your version? Latest is 104.10.1 . If the issue persists after restarting VS Code, please open a separate issue with detailed instructions.

**Comment by ryu-man** [comment]
Great, I tried that one more time and it's working fine now.

**Pull request merged**